### PR TITLE
Router package

### PR DESCRIPTION
**Pull request by Hyperkid123**
## New routing package 

Simple wrapper about stock `react-router-dom` components for easier migration to the shared history. Routing will still work if the application uses its own router, but the app will still use a separate history instance.

### TO DO
- [x] wrap all necessary components/hooks to automatically use router basename
- [x] tests

### CODE REVIEW

**Review by karelhala — APPROVED** [review]

File: packages/router/package.json [hunk 1]
```diff
@@ -0,0 +1,33 @@
+{
+    "name": "@redhat-cloud-services/router",
```
[the remark this answers is not in the document]
Comment (Hyperkid123) [reply packages/router/package.json:2]: Yeeeeeah we can do it. I guess. I will rename it after I am done with implementation and testing. its just soo long though :D

**Comment by Hyperkid123** [comment]
@karelhala I think I am done. I can't think of anything else that is missing. Once merged I would like to experiment with turning on the singleton flag in chrome in ci beta and see if it has any effects on existing application routing.

File: packages/router/src/Switch/index.js [hunk 2]
```diff
@@ -0,0 +1,3 @@
+export * from './Switch';
+export { default } from './Switch';
+
```
Comment (rvsia) [inline packages/router/src/Switch/index.js:3]: empty space

File: packages/router/src/useHistory/useHistory.js [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+        return internalHistory.replace(ignoreBasename ? path : internalPrependTo(path), state);
+    };
+
+    return {
+        ...internalHistory,
+        location,
+        push,
+        replace
+    };
+
```
Comment (rvsia) [inline packages/router/src/useHistory/useHistory.js:30]: empty line

File: packages/router/src/useHistory/useHistory.js [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    const internalHistory = routerUseHistory();
+    let internalRemovePathPrefix = typeof removePathPrefix === 'function' ? removePathPrefix : path => path;
+    let internalPrependTo = typeof prependTo === 'function' ? prependTo : path => path;
+
+    let location = {
+        ...internalHistory.location,
+        pathname: internalRemovePathPrefix(internalHistory.location.pathname)
+    };
+
+    const push = (path, state, ignoreBasename = false) => {
```
Comment (rvsia) [inline packages/router/src/useHistory/useHistory.js:16]: unnecessary return block

File: packages/router/src/useHistory/useHistory.js [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+        return internalHistory.push(ignoreBasename ? path : internalPrependTo(path), state);
+    };
+
+    const replace = (path, state, ignoreBasename = false) => {
+        return internalHistory.replace(ignoreBasename ? path : internalPrependTo(path), state);
```
Comment (rvsia) [inline packages/router/src/useHistory/useHistory.js:21]: ^

File: packages/router/src/useHistory/useHistory.test.js [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+
+const MatchComponent = ({ pushParams, method, expectedHistory, done }) => {
+    const history = useHistory();
+    const [ initialRender, setInitialRender ] = useState(false);
+    useEffect(() => {
+        if (!initialRender) {
+            history[method](...pushParams);
+            setInitialRender(true);
+        } else {
+            expect(history).toEqual(expectedHistory);
```
Comment (rvsia) [inline packages/router/src/useHistory/useHistory.test.js:33]: `assertions`

File: packages/router/src/useLocation/useLocation.js [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+    const internalLocation = routerUseLocation();
+    let internalRemovePathPrefix = typeof removePathPrefix === 'function' ? removePathPrefix : path => path;
+
+    let location = {
+        ...internalLocation,
+        pathname: internalRemovePathPrefix(internalLocation.pathname)
+    };
+
+    return location;
+
```
Comment (rvsia) [inline packages/router/src/useLocation/useLocation.js:16]: empty ,ine

File: packages/router/src/useLocation/useLocation.test.js [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+describe('useLocation', () => {
+    it('should return original location values', (done) => {
+        const pathname = '/foo/bar/1';
+        const expectedLocation = expect.objectContaining({
+            pathname
+        });
+        const MatchComponent = () => {
+            const location = useLocation();
+            useEffect(() => {
+                expect(location).toEqual(expectedLocation);
```
Comment (rvsia) [inline packages/router/src/useLocation/useLocation.test.js:33]: assertions

File: packages/router/src/useRouteMatch/useRouteMatch.js [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+    if (match) {
+        return {
+            ...match,
+            path: internalRemovePathPrefix(match.path),
+            url: internalRemovePathPrefix(match.url)
+        };
+    }
+
+    return match;
+
```
Comment (rvsia) [inline packages/router/src/useRouteMatch/useRouteMatch.js:26]: empty line

File: packages/router/src/useRouteMatch/useRouteMatch.test.js [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+            params: {
+                id: '1'
+            },
+            path,
+            url: '/foo/bar/1'
+        };
+        const MatchComponent = () => {
+            const match = useRouteMatch(path);
+            useEffect(() => {
+                expect(match).toEqual(expectMatch);
```
Comment (rvsia) [inline packages/router/src/useRouteMatch/useRouteMatch.test.js:38]: assertions

**Comment by rvsia** [comment]
^ just small nitpicks

**Review by rvsia — APPROVED** [review]
🏆

**Review by karelhala — APPROVED** [review]

File: packages/router/src/ChromeRouter/ChromeRouter.js [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+    }
+
+    return to;
+}
+
+function removePathPrefix(prefix = '', path = '') {
+    return path.replace(new RegExp(`^${prefix}/?`), '/');
+}
+
+const ChromeRouter = ({ basename = '/', children }) => (
```
[the remark this answers is not in the document]
Comment (Hyperkid123) [reply packages/router/src/ChromeRouter/ChromeRouter.js:28]: If we spread custom values you override the context value and the whole thing breaks.

File: packages/router/src/index.js [hunk 12]
```diff
@@ -0,0 +1,10 @@
+export * from 'react-router-dom';
+export { default as ChromeRouter } from './ChromeRouter';
+export { default as Route } from './Route';
+export { default as Link } from './Link';
+export { default as NavLink } from './NavLink';
+export { default as Redirect } from './Redirect';
+export { default as Switch } from './Switch';
+export { default as useLocation } from './useLocation';
+export { default as useHistory } from './useHistory';
+export { default as useRouteMatch } from './useRouteMatch';
```
[the remark this answers is not in the document]
Comment (Hyperkid123) [reply packages/router/src/index.js:10]: Rather there is no reason to expand it. It only holds functions that are used by the wrappers. There is no added value in using it directly.

**Pull request merged**